### PR TITLE
Add github ci to automate tests and to automate updating release numbers and add a code of conduct.

### PR DESCRIPTION
**Pull request by neon-mmd**
I have added `github ci` to `automate testing` of your project and added the ability to `auto update release numbers (versions)` by using [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) and also made writing conventional commits easier as well to get it setup on your system just run the following command:

``` shell
cargo install cc-cli
```

Usage information can be found [here](https://crates.io/crates/cc-cli) (which tends to be `pretty simple, easy and straight forward`).

Also I added a `code of conduct` to make contributing/collabrating to your project `much safer` for other developers.

On more information on how `automating release can help` check out this [video](https://piped.kavin.rocks/watch?v=fcHJZ4pMzBs) :smile:.

### CODE REVIEW

**Comment by maxcurzi** [comment]
@neon-mmd thanks for taking the time to work on this.

This PR does too many things (3, in fact). It would be best to create a PR for each "issue/feature", if your title is 
"Add XXX and YYY and ZZZ" it's a sign that the PR should be split.
This is because merging a PR is an all-or-nothing action; while one part may be ok, others may need some work.
Before you do that, please check my notes on this PR.

## Notes

### GITHUB CI
I had Github CI to automate tests, but as you may have noticed it's not just a matter of adding a Rust.yml file. The environment needs more setup or the task will fail (i.e. the runner must install various required dependencies). Since the task failed on your working branch (see screenshot), this part is not "done" and shouldn't be merged. 
image
From the log it looks like alsa is missing so probably the CI worker needs to install these requirements (as seen in the REAME.md) before running the tests: `libopencv-dev clang libclang-dev ffmpeg libavfilter-dev libavdevice-dev libasound2-dev yt-dlp`.
However, installing all prerequisites takes a very long time and it's not something you will want to do at every commit (it's energy wasteful, and it delays the near-instant feedback you'd want from unit tests).
Maybe there's a better way to set this up like creating a docker container once and having GI Actions to use this container rather than a generic one. I have never done that but in principle, it should work. It will take a bit of effort though.

### Automating updating release numbers
This is not something I want to have right now. I can see the merit of conventional commits but I'm not convinced right now.
1) To work well you'd want to install that `cc-cli` utility, this is fine if you just do commits from a computer. Sometimes I'd do commits from the web or mobile, and I may not remember the exact syntax for conventional commits. It will be easy to make mistakes. 
2) While automating releases sounds like a good idea, I would not trust it (due to the issue just mentioned). I like the idea that if a commit is tagged as "breaking change" the release will follow the correct SemVer convention, but for a project of this scope and size I still prefer doing a manual release.

Adding automation is not a zero-cost option, there is a setup/maintenance/troubleshooting burden associated with it. At the moment (and for the foreseeable future) the cost/effort of manually releasing is so little that I don't want to embark on this automation journey. Also, GIT releases so far are tied to a Crate release on crates.io and this is still something I want to have control on.
Therefore, adding automation will actually increase the work and effort I'd have to put in, rather than simplify my life.

### Code of conduct
I don't think a code of conduct is useful. This is not a Discord server :)
Contributing and collaborating to my project is already safe.
I collaborators/contributors are disrespectful/trolls etc.. etc... I will take measures, it's not like a Code of Conduct will deter trolls ;)


## In summary
1) Once you manage to get GH CI automated test to run successfully (and if it can run in ~1 minute), create a PR for it. 
2) There's no need for the other things (release automation, conventional commits, code of conduct). Maybe in the future... if the project continues to grow in scope/size and contributors/base.

**Comment by neon-mmd** [comment]
Ok sure, I will next time keep in mind to split multiple fixes and changes into separate PRs, and also maybe once I find a way to reduce the time it takes for the CI job to run I will add a new PR for it, right I have managed to reduce CI job time by 2 minutes by using caching but I think it is still too long so once I find a solution, I will be here again :smile: .
